services/fileuploader/fileuploader.go

```diff
@@ -121,13 +157,30 @@ func (p *provider) updateLoop(ctx context.Context, backendConfig backendconfig.B
 			if bucket.Type != "" && len(bucket.Config) > 0 {
 				preferences = c.Settings.DataRetention.StoragePreferences
 			}
-			settings[workspaceId] = StorageSettings{
+
+			// if no change in storage configuration, don't create new Filemanager
+			if ok && reflect.DeepEqual(currentWorkspaceConfig.Bucket, bucket) && reflect.DeepEqual(currentWorkspaceConfig.Preferences, preferences) {
+				settingsMap[workspaceId] = currentWorkspaceConfig
+				filemanagerMap[workspaceId] = currentFileManagerMap[workspaceId]
+				continue
+			}
+
+			// either newly polled workspace settings or updated storage config - update object storage Filemanager
+			settingsMap[workspaceId] = StorageSettings{
 				Bucket:      bucket,
 				Preferences: preferences,
+				updatedAt:   time.Now(),
 			}
+			filemanagerMap[workspaceId] = sync.OnceValues(func() (filemanager.FileManager, error) {
+				return filemanager.New(&filemanager.Settings{
+					Provider: bucket.Type,
+					Config:   bucket.Config,
+				})
+			})
 		}
 		p.mu.Lock()
-		p.storageSettings = settings
+		p.storageSettings = settingsMap
+		p.fileManagerMap = filemanagerMap
 		p.mu.Unlock()
 		p.onceInit.Do(func() {
 			close(p.init)
```

Comment:
Is the second `onceInit` after the for loop needed?

Reply:
I think it was added in the past to cater to some corner case, can't recall the reason though.
No harm in having it though right..?

Reply:
I don't think we should have "just in case" code. I think we should try our best to fully understand the code we are working on.

That said, I think this was done to cover the case where `ch` would get closed right away by `backendConfig`. This means that the `updateLoop` would never get to do an itearation of the `for` loop, thus it would never initialize things.

And if you don't "initialize" things then calls to, say, `GetStoragePreferences` would block forever. Now that we have a context, they would block until the context gets cancelled but in some places we're still passing a background one.

All considered, I think the correct way of doing it though is not to hackily say that we initialized things even if we didn't. I think we that if we are not subscribed to backend config changes, then we should say so with an error.

Perhaps along with the `init` channel, we can have a new one called `notSubscribed` and then do this towards the end of the `updateLoop`:

```go
		p.onceInit.Do(func() {
			close(p.init)
		})
	}

	p.onceInit.Do(func() {
		close(p.notSubscribed)
	})
```

Then we can change the other methods accordingly:

```go
func (p *provider) GetStoragePreferences(ctx context.Context, workspaceID string) (backendconfig.StoragePreferences, error) {
	var prefs backendconfig.StoragePreferences
	select {
	case <-ctx.Done():
		return prefs, ctx.Err()
	case <-p.init:
	case <-p.notSubscribed:
		return ErrNotSubscribed
	}
```

As things are at the moment, we're simply firing the `updateLoop` like this:

```go
go s.updateLoop(ctx, config)
```

If something goes wrong in the backend config then would we even notice?
* if we managed to init at least once we could serve stale settings until the service runs and for some workspaces we could return that there are not settings (if they came up afterwards)
* if we didn't manage to init ever, we wouldn't know and we would keep returning `ErrNoStorageForWorkspace` for example

WDYT?

Reply:
I understood the case in point, and the idea makes sense.

With the mentioned select scenario above though, it's possible that we still return stale configuration when config subscription stops(it could randomly chose between `<-p.init` and `<-p.notSubscribed`). Instead we could use a bool(atomic?) to return early `if !subscribed`.

```
func (p *provider) GetStoragePreferences(ctx context.Context, workspaceID string) (backendconfig.StoragePreferences, error) {
	var prefs backendconfig.StoragePreferences
	if !p.subscribed.Load() {
		return prefs, ErrNotSubscribed
	}
	select {
	case <-ctx.Done():
		return prefs, ctx.Err()
	case <-p.init:
	}
```
And
```
func (p *provider) updateLoop(ctx context.Context, backendConfig backendconfig.BackendConfig) {
	ch := backendConfig.Subscribe(ctx, backendconfig.TopicBackendConfig)
	defer p.subscribed.Load(false)
	for ev := range ch {
		// process config
		p.once.Do(func() {
			close(p.init)
			p.subscribed.Load(true)
		})
	}
}

Reply:
That could work. The alternative would be to have two `select`s as to specify the priority. Your approach seems leaner though considering that there is no way to be always 100% precise and consistent here. So +1 for the atomic value :slightly_smiling_face:

Reply:
Removed the `Once` and the `init` channel too. The bool sufficed.

Reply:
I've added a draft [here](https://github.com/rudderlabs/rudder-server/pull/4989).